pilot/cmd/pilot-discovery/server/server.go

```diff
@@ -0,0 +1,391 @@
+// load the mesh from the file. Otherwise, a default mesh will be used with optional overrides.
+type MeshArgs struct {
+	ConfigFile      string
+	MixerAddress    string
+	RdsRefreshDelay *durpb.Duration
+}
+
+// ConfigArgs provide configuration options for the configuration controller. If FileDir is set, that directory will
+// be monitored for CRD yaml files and will update the controller as those files change (This is used for testing
+// purposes). Otherwise, a CRD client is created based on the configuration.
+type ConfigArgs struct {
+	KubeConfig        string
+	ControllerOptions kube.ControllerOptions
+	FileDir           string
+}
+
+// ConsulArgs provides configuration for the Consul service registry.
+type ConsulArgs struct {
+	Config    string
+	ServerURL string
+}
+
+// EurekaArgs provides configuration for the Eureka service registry
+type EurekaArgs struct {
+	ServerURL string
+}
+
+// ServiceArgs provides the composite configuration for all service registries in the system.
+type ServiceArgs struct {
+	Registries []string
+	Consul     ConsulArgs
+	Eureka     EurekaArgs
+}
+
+// PilotArgs provides all of the configuration parameters for the Pilot discovery service.
+type PilotArgs struct {
+	DiscoveryOptions envoy.DiscoveryServiceOptions
+	Namespace        string
+	Mesh             MeshArgs
+	Config           ConfigArgs
+	Service          ServiceArgs
+}
+
+// Server contains the runtime configuration for the Pilot discovery service.
+type Server struct {
+	config *pilotConfig
+}
+
+// NewServer creates a new Server instance based on the provided arguments.
+func NewServer(args PilotArgs, stop chan struct{}) (*Server, error) {
+	config := &pilotConfig{stop: stop}
+
+	// Apply the arguments to the configuration.
+	if err := aggregateConfigFn(&args, config); err != nil {
+		return nil, multierror.Prefix(err, "failed to create Pilot discovery service.")
+	}
+
+	return &Server{
+		config: config,
+	}, nil
+}
+
+// Run starts the pilot discovery service.
+func (s Server) Run() {
+	// Now start all of the services.
+	for _, runFn := range s.config.runFuncs {
+		runFn()
+	}
+}
+
+// runFunc is a no-op function that will be executed after the Pilot discovery service has been configured.
+type runFunc func()
+
+// pilotConfig contains the runtime structures for the Pilot discovery service. This structure is configured
+// by a number of pilotConfigFunc functions.
+type pilotConfig struct {
+	stop              chan struct{}
+	mesh              *meshconfig.MeshConfig
+	serviceController *aggregate.Controller
+	configController  model.ConfigStoreCache
+	mixerSAN          []string
+	runFuncs          []runFunc
+}
+
+// pilotConfigFunc is a function that configures some part of the pilotConfig based on the settings in the PilotArgs.
+type pilotConfigFunc func(in *PilotArgs, out *pilotConfig) error
+
+// initMesh creates the mesh in the pilotConfig from the input arguments.
+func initMesh(in *PilotArgs, out *pilotConfig) error {
```

Comment:
Can these args be read from a file?

Reply:
I don't see why not ... are you proposing that be added to this PR? This particular file doesn't care much about the origin of the args.

Reply:
nm. dumb question.. Once we move global mesh config to CRDs, we can reuse the same file adapter you have today